src/apps/engine/src/compiler.cpp

```diff
@@ -7117,6 +7273,509 @@ void COMPILER::PrintoutUsage()
+        {
+            auto vi = VarTab.GetVarX(relocated_data);
+            variables.emplace(relocated_data, vi->name);
+        }
+        else if (token_type == CALL_FUNCTION)
+        {
+            auto fi = FuncInfo();
+            FuncTab.GetFuncX(fi, relocated_data);
+            functions.emplace(relocated_data, fi.name);
+        }
+    } while (token_type != END_OF_PROGRAMM);
+
+    writer.WriteData(strings.size());
+    for (auto &[string_code, string] : strings)
+    {
+        writer.WriteData(string_code);
+        writer.WriteBytes(string);
+    }
+
+    writer.WriteData(variables.size());
+    for (auto &[variable_code, name] : variables)
+    {
+        writer.WriteData(variable_code);
+        writer.WriteBytes(name);
+    }
+
+    writer.WriteData(functions.size());
+    for (auto &[function_code, name] : functions)
+    {
+        writer.WriteData(function_code);
+        writer.WriteBytes(name);
+    }
+}
+
+void COMPILER::SaveSegmentToCache(const SEGMENT_DESC &segment)
+{
+    const auto path = GetSegmentCachePath(segment);
+    create_directories(path.parent_path());
+    auto stream = std::ofstream(path, std::ios::binary);
+
+    if (!stream.is_open())
+    {
+        return;
+    }
+
+    auto writer = storm::script_cache::Writer();
+
+    // save files lists and total crc
+    SaveFilesToCache(writer);
+
+    // defines (in case of new compiled code depending on defines from cache)
+    SaveDefinesToCache(writer);
+
+    // variables with values
+    SaveVariablesToCache(writer);
+
+    // functions
+    SaveFunctionsToCache(writer);
+
+    // script libs
+    SaveScriptLibrariesToCache(writer);
+
+    // event handlers
+    SaveEventHandlersToCache(writer);
+
+    // bytecode
+    SaveByteCodeToCache(writer, segment);
+
+    auto &data = writer.GetData();
+    stream.write(data.data(), data.size());
+}
+
+void COMPILER::SaveFilesToCache(storm::script_cache::Writer &writer)
+{
+    // save crc for verification
+    writer.WriteData(script_cache_.crc);
+
+    writer.WriteData(script_cache_.files.size());
+    for (auto &file_name : script_cache_.files)
+    {
+        writer.WriteBytes(file_name);
+    }
+}
+
+void COMPILER::SaveDefinesToCache(storm::script_cache::Writer &writer)
+{
+    writer.WriteData(script_cache_.defines.size());
+    for (auto &[name, type, value] : script_cache_.defines)
+    {
+        writer.WriteBytes(name);
+        writer.WriteData(type);
+        switch (type)
+        {
+        case NUMBER:
+            writer.WriteData(static_cast<int32_t>(value));
+            break;
+
+        case FLOAT_NUMBER:
+            writer.WriteData(static_cast<float>(value));
+            break;
+
+        case STRING:
+            writer.WriteBytes(reinterpret_cast<const char*>(value));
+            break;
+        }
+    }
+}
+
+bool COMPILER::LoadSegmentFromCache(SEGMENT_DESC &segment)
+{
+    const auto path = GetSegmentCachePath(segment);
+    if (!exists(path))
```

Comment:
this may throw, please handle the exceptions

Reply:
done